File: scimesh/worker/daemon.py
"""The worker state machine and its safe failure handling."""

from __future__ import annotations

import logging
from dataclasses import replace
from pathlib import Path
import random
import shutil
import threading
import time
from datetime import datetime, timezone

from .artifacts import ArtifactClient, sha256_file
from .config import WorkerConfig
from .coordinator import CoordinatorClient, CoordinatorConflictError, CoordinatorTransientError
from .models import ClaimedTask, UploadedArtifact
from .runners import Runner


class LeaseHeartbeat:
    """Renews a claimed task lease while local work is in progress."""

    def __init__(self, task: ClaimedTask, coordinator: CoordinatorClient, config: WorkerConfig) -> None:
        self.task, self.coordinator, self.config = task, coordinator, config
        self._stop = threading.Event()
        self._error: Exception | None = None
        self._thread: threading.Thread | None = None
        self._lease_expires_at = task.lease_expires_at

    def start(self) -> None:
        # Verify ownership before expensive download or calculation begins.
        self._lease_expires_at = self.coordinator.heartbeat(
            self.task, self.config.worker_id
        )
        self._next_delay()
        self._thread = threading.Thread(target=self._run, name=f"lease-{self.task.task_id}", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread:
            self._thread.join()

    def raise_if_failed(self) -> None:
        if self._error:
            raise self._error

    def _run(self) -> None:
        delay = self._next_delay()
        while not self._stop.wait(max(delay, 0.01)):
            try:
                self._lease_expires_at = self.coordinator.heartbeat(
                    self.task, self.config.worker_id
                )
                delay = self._next_delay()
            except Exception as error:  # Surface the lease loss in the main state machine.
                self._error = error
                return

    def _next_delay(self) -> float:
        return min(self.config.heartbeat_interval, self._seconds_until_expiry() / 2)

    def _seconds_until_expiry(self) -> float:
        try:
            expiry = datetime.fromisoformat(self._lease_expires_at.replace("Z", "+00:00"))
        except ValueError as error:
            raise ValueError("invalid lease_expires_at") from error
        seconds = (expiry - datetime.now(timezone.utc)).total_seconds()
        if seconds <= 0:
            raise ValueError("claimed task lease has already expired")
        return seconds


class WorkerDaemon:
    def __init__(self, config: WorkerConfig, coordinator: CoordinatorClient, artifacts: ArtifactClient, runner: Runner) -> None:
        self.config, self.coordinator, self.artifacts, self.runner = config, coordinator, artifacts, runner
        self.worker_id = config.worker_id
        self._registered = False
        self.log = logging.getLogger("scimesh.worker")

    def run_forever(self) -> None:
        failures = 0
        processed_tasks = 0
        self._log(
            "started",
            max_tasks=self.config.max_tasks,
            exit_when_idle=self.config.exit_when_idle,
        )
        try:
            while True:
                try:
                    if not self._registered:
                        self._register_worker()
                    self._cleanup_expired_directories()
                    claimed = self.run_once()
                    failures = 0
                    if claimed:
                        processed_tasks += 1
                        if self.config.max_tasks is not None and processed_tasks >= self.config.max_tasks:
                            self._log("stopped", reason="max_tasks_reached", processed_tasks=processed_tasks)
                            return
                    elif self.config.exit_when_idle:
                        self._log("stopped", reason="queue_empty", processed_tasks=processed_tasks)
                        return
                    else:
                        self._sleep(self.config.poll_interval)
                except CoordinatorTransientError as error:
                    failures += 1
                    self._log("failed", error_type=type(error).__name__)
                    self._sleep(min(self.config.poll_interval * 2 ** min(failures, 6), 60.0))
        except KeyboardInterrupt:
            self._log("stopped", reason="interrupted", processed_tasks=processed_tasks)

    def run_once(self) -> bool:
        worker_id = self._worker_id()
        self._log("claiming", log_level=logging.DEBUG)
        task = self.coordinator.claim(worker_id, self.config.capabilities)
        if task is None:
            self._log("idle", log_level=logging.DEBUG)
            return False
        started = time.monotonic()
        task_dir = self.config.work_dir / task.task_id / str(task.attempt)
        heartbeat = LeaseHeartbeat(task, self.coordinator, self.config)
        try:
            task_dir.mkdir(parents=True, exist_ok=False)
            heartbeat.start()
            self._log("downloading", task)
            input_path = task_dir / "input"
            self.artifacts.download(task.input.uri, input_path)
            if sha256_file(input_path).lower() != task.input.sha256.lower():
                raise ValueError("input checksum mismatch")
            self._log("running", task)
            result = self.runner.run(task, task_dir)
            heartbeat.raise_if_failed()
            if len(result.artifacts) != 1:
                raise ValueError("runner must produce exactly one result artifact")
            artifact = result.artifacts[0]
            uploaded = self.artifacts.upload(task, worker_id, artifact)
            manifest = self._result_manifest(uploaded)
            self._log("submitting", task)
            heartbeat.raise_if_failed()
            self.coordinator.submit(
                task,
                {
                    "worker_id": worker_id,
                    "attempt": task.attempt,
                    "result": manifest,
                    "metrics": {
                        **result.metrics,
                        "elapsed_seconds": round(time.monotonic() - started, 3),
                    },
                },
            )
            self._log("completed", task, elapsed_seconds=round(time.monotonic() - started, 3))
        except CoordinatorConflictError as error:
            self._log("lease_lost", task, error_type=type(error).__name__)
        except Exception as error:
            self._log("failed", task, error_type=type(error).__name__)
            self._report_failure(task, error)
        finally:
            heartbeat.stop()
        return True

    def _report_failure(self, task: ClaimedTask, error: Exception) -> None:
        message = str(error).replace(str(self.config.work_dir), "<worker-dir>")[:300]
        try:
            self.coordinator.fail(task, {"worker_id": self._worker_id(), "attempt": task.attempt, "error_code": type(error).__name__, "error_message": message})
        except CoordinatorTransientError:
            raise
        except Exception:
            self._log("failed", task, error_type="FailureReportError")

    def _register_worker(self) -> None:
        registered = self.coordinator.register(
            self.config.worker_name,
            self.config.capabilities,
            self.config.cpu_count,
            self.config.memory_mb,
        )
        self.worker_id = registered.worker_id
        self.config = replace(
            self.config,
            worker_id=registered.worker_id,
            heartbeat_interval=registered.heartbeat_interval_seconds,
        )
        self._registered = True
        self._log("registered")

    def _worker_id(self) -> str:
        if not self.worker_id:
            raise ValueError("worker is not registered")
        return self.worker_id

    @staticmethod
    def _result_manifest(uploaded: UploadedArtifact) -> dict[str, object]:
        """Keep completion payload exact: coordinator owns all artifact metadata."""
        return {"artifact_id": uploaded.artifact_id}

    def _log(
        self,
        state: str,
        task: ClaimedTask | None = None,
        *,
        log_level: int = logging.INFO,
        **extra: object,
    ) -> None:
        fields = {"worker_id": self.config.worker_id, "task_id": task.task_id if task else None, "attempt": task.attempt if task else None, "state": state, **extra}
        self.log.log(log_level, "worker_event %s", fields)

    def _cleanup_expired_directories(self) -> None:
        """Remove only old task attempt directories when retention was configured."""
        if self.config.cleanup_after_seconds is None or not self.config.work_dir.exists():
            return
        cutoff = time.time() - self.config.cleanup_after_seconds
        for task_dir in self.config.work_dir.iterdir():
            if not task_dir.is_dir():
                continue
            for attempt_dir in task_dir.iterdir():
                if attempt_dir.is_dir() and attempt_dir.stat().st_mtime < cutoff:
                    shutil.rmtree(attempt_dir)
            if not any(task_dir.iterdir()):
                task_dir.rmdir()

    @staticmethod
    def _sleep(delay: float) -> None:
        time.sleep(delay * random.uniform(0.75, 1.25))

File: coordinator/internal/agent/daemon.go
package agent

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// Outcome reports whether a claim was made and whether it completed.
type Outcome struct {
	Claimed   bool
	Completed bool
}

// Daemon is the agent state machine: register, claim, execute via the Python
// task runner, upload, and submit — mirroring the Python worker's lifecycle.
type Daemon struct {
	config     *Config
	client     *Client
	runner     *TaskRunner
	log        *slog.Logger
	workerID   string
	registered bool
	completed  int
	mu         sync.Mutex
}

func NewDaemon(config *Config, client *Client, runner *TaskRunner, log *slog.Logger) *Daemon {
	return &Daemon{config: config, client: client, runner: runner, log: log}
}

// RunForever loops until interrupted, idle-exit, or max tasks.
func (d *Daemon) RunForever() error {
	failures := 0
	for {
		if !d.registered {
			if err := d.register(); err != nil {
				return err
			}
		}
		outcome, err := d.runOnce()
		if err != nil {
			failures++
			d.log.Warn("agent cycle failed", "error", err)
			backoff := d.config.PollInterval
			for i := 0; i < failures && i < 6; i++ {
				backoff *= 2
			}
			if backoff > 60*time.Second {
				backoff = 60 * time.Second
			}
			time.Sleep(backoff)
			continue
		}
		failures = 0
		if outcome.Claimed && outcome.Completed {
			d.completed++
			if d.config.MaxTasks > 0 && d.completed >= d.config.MaxTasks {
				d.log.Info("max tasks reached")
				return nil
			}
		}
		if !outcome.Claimed && d.config.ExitWhenIdle {
			d.log.Info("queue empty, exiting")
			return nil
		}
		if outcome.Claimed && d.config.ExitWhenIdle {
			d.log.Info("one claim processed, exiting")
			return nil
		}
		if !outcome.Claimed {
			time.Sleep(d.config.PollInterval)
		}
	}
}

func (d *Daemon) register() error {
	registered, err := d.client.Register(
		d.config.WorkerName,
		d.config.Capabilities,
		d.config.CPUCount,
		d.config.MemoryMB,
	)
	if err != nil {
		return err
	}
	d.mu.Lock()
	if d.config.WorkerID != "" {
		d.workerID = d.config.WorkerID
	} else {
		d.workerID = registered.WorkerID
	}
	d.registered = true
	d.mu.Unlock()
	d.log.Info("registered", "worker_id", d.workerID)
	return nil
}

func (d *Daemon) workerIDOrEmpty() string {
	d.mu.Lock()
	defer d.mu.Unlock()
	return d.workerID
}

func (d *Daemon) runOnce() (Outcome, error) {
	workerID := d.workerIDOrEmpty()
	if workerID == "" {
		return Outcome{}, fmt.Errorf("agent is not registered")
	}
	task, err := d.client.Claim(workerID, d.config.Capabilities)
	if err != nil {
		return Outcome{}, err
	}
	if task == nil {
		return Outcome{Claimed: false}, nil
	}
	started := time.Now()
	taskDir := filepath.Join(d.config.WorkDir, task.TaskID, fmt.Sprint(task.Attempt))
	if err := os.MkdirAll(taskDir, 0o755); err != nil {
		return Outcome{Claimed: true}, err
	}
	heartbeat := newLeaseHeartbeat(task, workerID, d.client, d.config.Heartbeat)
	completed := false
	err = heartbeat.Start()
	if err != nil {
		if _, ok := err.(*ConflictError); ok {
			d.log.Warn("lease lost", "task_id", task.TaskID)
			return Outcome{Claimed: true}, nil
		}
		return Outcome{Claimed: true}, err
	}
	defer heartbeat.Stop()

	// Attempt directory cleanup is deliberately minimal in the prototype:
	// attempt directories are retained under the work directory.
	failure := d.executeTask(task, workerID, taskDir, started, heartbeat)
	if failure != nil {
		if _, ok := failure.(*ConflictError); ok {
			d.log.Warn("lease lost", "task_id", task.TaskID)
			return Outcome{Claimed: true}, nil
		}
		if err := heartbeat.RaiseIfFailed(); err != nil {
			return Outcome{Claimed: true}, nil
		}
		d.reportFailure(task, workerID, failure)
		return Outcome{Claimed: true}, nil
	}
	if err := heartbeat.RaiseIfFailed(); err != nil {
		return Outcome{Claimed: true}, nil
	}
	completed = true
	d.log.Info("task completed", "task_id", task.TaskID, "elapsed_seconds", time.Since(started).Seconds())
	return Outcome{Claimed: true, Completed: completed}, nil
}

// executeTask returns nil on success or a classified failure.
func (d *Daemon) executeTask(task *Task, workerID, taskDir string, started time.Time, heartbeat *leaseHeartbeat) error {
	inputPath := filepath.Join(taskDir, "input")
	// Downloads use the coordinator-provided URI verbatim; a relative path is
	// resolved against the coordinator by the client.
	actualSHA, err := d.client.Download(task.Input.URI, inputPath)
	if err != nil {
		return err
	}
	if !strings.EqualFold(actualSHA, task.Input.SHA256) {
		return &CoordinatorError{msg: "input checksum mismatch"}
	}
	if err := heartbeat.RaiseIfFailed(); err != nil {
		return err
	}
	manifestPath := filepath.Join(taskDir, "manifest.json")
	manifest, err := d.runner.Run(task, taskDir, manifestPath, nil)
	if err != nil {
		return err
	}
	if err := heartbeat.RaiseIfFailed(); err != nil {
		return err
	}
	uploaded, err := d.client.Upload(task, workerID, manifest.ArtifactPath, manifest.ContentType)
	if err != nil {
		return err
	}
	if err := heartbeat.RaiseIfFailed(); err != nil {
		return err
	}
	metrics := map[string]any{"elapsed_seconds": roundSeconds(time.Since(started).Seconds())}
	for name, value := range manifest.Metrics {
		metrics[name] = value
	}
	return d.client.Submit(task, workerID, uploaded, metrics)
}

func (d *Daemon) reportFailure(task *Task, workerID string, failure error) {
	var code string
	switch failure.(type) {
	case *CoordinatorError:
		code = "ValueError"
	default:
		code = "TaskRunnerFailed"
	}
	retryable := IsRetryableError(failure)
	message := SanitizeErrorMessage(failure.Error(), d.config.WorkDir)
	d.log.Warn("task failed", "task_id", task.TaskID, "error_code", code, "retryable", retryable)
	if err := d.client.Fail(task, workerID, code, message, retryable); err != nil {
		if _, ok := err.(*ConflictError); ok {
			d.log.Warn("lease lost while reporting failure", "task_id", task.TaskID)
			return
		}
		d.log.Warn("failure report rejected", "task_id", task.TaskID, "error", err)
	}
}

// leaseHeartbeat renews the lease from the returned deadline at less than
// half of the remaining TTL, mirroring the Python worker.
type leaseHeartbeat struct {
	task     *Task
	workerID string
	client   *Client
	interval time.Duration
	stop     chan struct{}
	once     sync.Once
	mu       sync.Mutex
	lease    time.Time
	failed   error
}

func newLeaseHeartbeat(task *Task, workerID string, client *Client, interval time.Duration) *leaseHeartbeat {
	return &leaseHeartbeat{
		task:     task,
		workerID: workerID,
		client:   client,
		interval: interval,
		stop:     make(chan struct{}),
		lease:    task.LeaseExpiresAt,
	}
}

func (h *leaseHeartbeat) Start() error {
	if _, err := h.client.Heartbeat(h.task, h.workerID); err != nil {
		return err
	}
	go h.loop()
	return nil
}

func (h *leaseHeartbeat) Stop() {
	h.once.Do(func() { close(h.stop) })
}

func (h *leaseHeartbeat) RaiseIfFailed() error {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.failed
}

func (h *leaseHeartbeat) loop() {
	for {
		delay := h.nextDelay()
		select {
		case <-h.stop:
			return
		case <-time.After(delay):
		}
		renewed, err := h.client.Heartbeat(h.task, h.workerID)
		h.mu.Lock()
		if err != nil {
			h.failed = err
			h.mu.Unlock()
			return
		}
		h.lease = renewed
		h.mu.Unlock()
	}
}

func (h *leaseHeartbeat) nextDelay() time.Duration {
	h.mu.Lock()
	defer h.mu.Unlock()
	remaining := time.Until(h.lease)
	if remaining <= 0 {
		return 0
	}
	half := remaining / 2
	if h.interval < half {
		return h.interval
	}
	return half
}

func roundSeconds(seconds float64) float64 {
	return float64(int64(seconds*1000)) / 1000
}

// File-digest helper used by tests.
func sha256File(path string) (string, error) {
	file, err := os.Open(path)
	if err != nil {
		return "", err
	}
	defer file.Close()
	digest := sha256.New()
	if _, err := io.Copy(digest, file); err != nil {
		return "", err
	}
	return hex.EncodeToString(digest.Sum(nil)), nil
}
